### PR TITLE
CRS-216 MessageInput

### PR DESCRIPTION
**Pull request by ath92**
## Description of the pull request
MessageInput using hooks

Still missing:
 - TS definitions need updating + comments need to be added
 - Test branch was merged into this, so will need to clean up git history later

### CODE REVIEW

**Comment by vicnicius** [comment]
Except for some types still failing, LGTM 👍

File: src/components/MessageInput/MessageInputState.js [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+        fileUploads: state.fileUploads.without(action.id),
+      };
+    case 'reduceNumberOfUploads': // TODO: figure out if we can just use uploadOrder instead
+      return { ...state, numberOfUploads: state.numberOfUploads - 1 };
+    case 'addMentionedUser':
+      return {
+        ...state,
+        mentioned_users: state.mentioned_users.concat(action.userId),
+      };
+    default:
```
Comment (vicnicius) [inline src/components/MessageInput/MessageInputState.js:217]: Maybe we could add a warning if the reducer defaults.

**Comment by ath92** [comment]
Yeah the types failing should be resolved once my commit to the js client is merged :)

File: src/components/MessageInput/MessageInputState.js [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+ *  type: 'reduceNumberOfUploads',
+ * }} ReduceNumberOfUploadsAction
+ * @typedef {{
+ *  type: 'addMentionedUser',
+ *  userId: string,
+ * }} AddMentionedUserAction
+ *
+ * MessageInput state reducer
+ * @param {State} state
+ * @param {SetEmojiPickerIsOpenAction | SetTextAction | ClearAction | SetImageUploadAction | SetFileUploadAction | RemoveImageUploadAction | RemoveFileUploadAction | ReduceNumberOfUploadsAction | AddMentionedUserAction} action
```
Comment (vicnicius) [inline src/components/MessageInput/MessageInputState.js:147]: Maybe define a separate type for those actions and use it here:
```typescript
/**
 * @typedef {
 * | SetEmojiPickerIsOpenAction
 * | SetTextAction
 * | ClearAction
 * | SetImageUploadAction
 * | SetFileUploadAction
 * | RemoveImageUploadAction
 * | RemoveFileUploadAction
 * | ReduceNumberOfUploadsAction
 * | AddMentionedUserAction
 * } MessageInputStateAction
 *
 * MessageInput state reducer
 * @param {State} state
 * @param {MessageInputStateAction} action
 * @returns {State}
 */
```

File: src/components/MessageInput/MessageInputTypes.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
+type SetEmojiPickerIsOpenAction = {
```
Comment (vicnicius) [inline src/components/MessageInput/MessageInputTypes.ts:1]: I definitely think that it is way better to extract the types to their own files like this. 👍 I only wonder if we should go with one per component or if we could have one type file per "namespace": Like `src/components/MessageInput/types.ts`. But that's something we can discuss and adjust along the way.

Comment (ath92) [reply src/components/MessageInput/MessageInputTypes.ts:1]: Ah yes that also makes sense, I wasn't really sure what to call this so if we just go with types.ts, I won't have to care about naming anymore :)

**Pull request merged**